pennylane/devices/preprocess.py

```diff
@@ -545,3 +544,211 @@ def _validate_snapshot_shots(tape, sample_measurements, analytic_measurements, n
+
+    .. note::
+        This transform diagonalizes all the operations on the tape. An error will
+        be raised if non-commuting terms are encountered. To avoid non-commuting
+        terms in circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>`
+        transform can be applied.
+
+    Args:
+        tape (QNode or QuantumTape or Callable): A quantum circuit.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], function]: The
+        transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    **Example**
+
+    Consider the tape:
+
+    >>> ops = [qml.X(0), qml.RY(1.23, 1)]
+    >>> measurements = [qml.expval(qml.Y(0)), qml.probs(wires=[1])]
+    >>> tape = qml.tape.QuantumScript(ops, measurements, shots=10)
+
+    We can apply the transform to diagonalize and convert the two measurements to a single sample:
+
+    >>> (new_tape, ), fn = qml.transforms.measurements_from_samples(tape)
+    >>> new_tape.measurements
+    [sample(wires=[0, 1])]
+
+    The tape operations now include diagonalizing gates.
+
+    >>> new_tape.operations
+    [X(0), RY(1.23, wires=[1]), RX(1.5707963267948966, wires=[0])]
+
+    Executing the tape returns samples that can be post-processed to get the originally requested measurements:
+
+    >>> dev = qml.device("default.qubit")
+    >>> dev.execute(new_tape)
+    array([[1, 0],
+           [0, 0],
+           [0, 1],
+           [1, 1],
+           [0, 1],
+           [1, 0],
+           [0, 1],
+           [1, 0],
+           [1, 0],
+           [1, 0]])
+
+    >>> fn((res,))
+    (-0.2, array([0.6, 0.4]))
+    """
+    if tape.shots.total_shots is None:
+        return (tape,), null_postprocessing
+
+    for mp in tape.measurements:
+        if not mp.obs and not mp.wires:
+            raise RuntimeError(
+                "Please apply validate_device_wires transform before measurements_from_samples"
+            )
+
+    diagonalized_tape, measured_wires = _get_diagonalized_tape_and_wires(tape)
+    new_tape = diagonalized_tape.copy(measurements=[qml.sample(wires=measured_wires)])
+
+    def unsqueezed(samples):
+        """If the samples have been squeezed to remove the 'extra' dimension in the case where
+        shots=1 or wires=1, unsqueeze to restore the raw samples format expected by mp.process_samples
+        """
+
+        # if we stop squeezing out the extra dimension for shots=1 or wires=1 when sampling (as is
+        # already the case in Catalyst), this problem goes away
+
+        if len(samples.shape) == 1:
+            samples = qml.math.array([[s] for s in samples], like=samples)
+        return samples
+
+    def postprocessing_fn(results):
+        """A processing function to get measurement values from samples."""
+        samples = results[0]
+        results_processed = []
+        for m in tape.measurements:
+            if len(tape.shots.shot_vector) > 1:
```

Comment:
```suggestion
            if tape.shots.has_partitioned_shots:
```

```
>>> len(qml.measurements.Shots((100, 100)).shot_vector)
1
```
The `shot_vector` is a tuple of `ShotCopies`, which is a number of shots a certain number of times.  For example 100 2 times.